wemake_python_styleguide/visitors/ast/statements.py

```diff
@@ -498,3 +508,108 @@ def _check_tuple_arguments_types(
             if isinstance(arg, self._no_tuples_collections):
                 self.add_violation(NotATupleArgumentViolation(node))
                 break
+
+
+@final
+class UnsafeGeneratorExpressionVisitor(BaseNodeVisitor):
+    """Checks if a generator expression is safe."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Creates the lists that would hold the variables."""
+        super().__init__(*args, **kwargs)
+        self.inside_variables: List[str] = []
+        self.created_variables: List[str] = []
+        self.affecting_variables: List[str] = []
+
+    def visit(self, node: ast.AST) -> None:
+        """Checks if the generator expression has affecting variables."""
+        assignments = [
+            sub_node
+            for sub_node in ast.walk(node)
+            if isinstance(sub_node, ast.Assign)
+        ]
+        generators = [
+            generator.value
+            for generator in assignments
+            if isinstance(generator.value, ast.GeneratorExp)
+        ]
+
+        if generators:
+            for generator in generators:
+                self._get_inside_variables(generator)
+                self._check_generator_variables(generator)
+                
+                inside = self.inside_variables
+                created = self.created_variables
+
+                self.affecting_variables = ([
+                    name
+                    for name in inside
+                    if name not in inside or name not in created
+                ])
+
+                if self.affecting_variables:
+                    self._check_gen_usage(node, assignments)
+
+    def _get_inside_variables(self, node: ast.GeneratorExp) -> None:
+        sub_node = node.elt
+        self.inside_variables = [
```

Comment:
We already have this logic as a function: https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/logic/walk.py#L56